### PR TITLE
GSL UPP developments up to January 2023

### LINKED ISSUE #610: Resolve remaining differences with GSL RRFS_dev UPP
This change will make the GSL RRFS_dev UPP identical to the authoritative develop branch, with the exception of IFI.  Changes include:

- A change to GRIB2 precision for VIL (it was previously insufficient)
- A change to GSL's experimental ceiling diagnostic #2.
- Resolve difference in name of cldfra variable between EMC and GSL RRFS runs.

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
1. Increases precision of VIL_ON_ENTIRE_ATMOS from `<scale>4.0</scale>` to `<scale>7.0</scale>`
2. Adds an experimental diagnostic, developed for the HRRR to correct for the HRRR's low bias in cloud cover. For an FAA deliverable due Feb 2023.
3. Able to use either cldfra or cldfra_bl when reading cloud fraction from INITPPOST_NETCDF. (Will use whichever is available; uses cldfra if both are present.)

The GSL in-flight icing developments were merged in pull requests #536 and #621

Fixes #610

### CODE REVIEW

**Comment by SamuelTrahanNOAA** [comment]
These changes were from @EricJames-NOAA and @hu5970

**Comment by EricJames-NOAA** [comment]
@SamuelTrahanNOAA thanks!  

Regarding change (3), I don't understand why EMC reads in a different field for cloud fraction from FV3.  GSL has always had cldfra_bl output.  We may need to ask other developers about this.

**Comment by WenMeng-NOAA** [comment]
> @SamuelTrahanNOAA thanks!
> 
> Regarding change (3), I don't understand why EMC reads in a different field for cloud fraction from FV3. GSL has always had cldfra_bl output. We may need to ask other developers about this.

@yangfanglin and @ericaligo-NOAA Could you confirm if EMC side has switched cloud fraction model output from cldfra to cldfra_bl for RRFS?

**Comment by ericaligo-NOAA** [comment]
I added the universal 3D cloud fraction,cldfra, in 2020 I believe in order for it to be used in all microphysics applications including the GFDL microphysics. There was no other instantaneous cloud fraction in the FV3 output at that time.Eric
-------- Original message --------From: WenMeng-NOAA ***@***.***> Date: 1/26/23  9:02 PM  (GMT-05:00) To: NOAA-EMC/UPP ***@***.***> Cc: ericaligo-NOAA ***@***.***>, Mention ***@***.***> Subject: Re: [NOAA-EMC/UPP] GSL UPP developments up to January 2023 (PR #620) 

@SamuelTrahanNOAA thanks!
Regarding change (3), I don't understand why EMC reads in a different field for cloud fraction from FV3. GSL has always had cldfra_bl output. We may need to ask other developers about this.

@yangfanglin and @ericaligo-NOAA Could you confirm if EMC side has switched cloud fraction model output from cldfra to cldfra_bl for RRFS?

**Comment by SamuelTrahanNOAA** [comment]
~~I've merged the bug fix from #621 that is needed to get the inline post working again, after my libIFI changes broke it. (I don't deallocate some arrays after the next iteration of the post.) If this PR will take a while to review, then we should merge #621 first since it'll be a shoo-in.~~

EDIT: The bug fix in #621 has been merged.

**Comment by SamuelTrahanNOAA** [comment]
The bug fix in #621 has been merged to develop, and I've synced this branch with develop again.

File: sorc/ncep_post.fd/IFI.F [hunk 1]
```diff
@@ -199,6 +199,7 @@ subroutine make_communicators
     ! and for determining root ranks.
     allocate(ista_grid(size))
     allocate(jsta_grid(size))
+
```
Comment (SamuelTrahanNOAA) [inline sorc/ncep_post.fd/IFI.F]:
```suggestion
```
Remove accidental blank line. This came from the inline post bug fix. It was removed from that PR, but hasn't been removed from this one yet.

File: sorc/ncep_post.fd/INITPOST_NETCDF.f [hunk 2]
```diff
@@ -889,7 +889,7 @@ SUBROUTINE INITPOST_NETCDF(ncid2d,ncid3d)
         call read_netcdf_3d_para(ncid3d,im,jm,ista,ista_2l,iend,iend_2u,jsta,jsta_2l,jend,jend_2u, &
         spval,VarName,cfr(ista_2l,jsta_2l,1),lm)
       else
-        VarName='cldfra'
+        VarName='cldfra_bl'
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/INITPOST_NETCDF.f]: @SamuelTrahanNOAA It appears that reading "cldfra" is needed by EMC side. I would suggest adding changes in the subroutine read_netcdf_2d_para for reading "cldfra_bl" when "cldfra" is missing in model output, such as:
```
iret = nf90_inq_varid(ncid,trim(varname),varid)
if (iret == 0 .and. trim(varname) == "cldfra") then
   iret = nf90_inq_varid(ncid,"cldfra_bl",varia) !!reading cldfra_bl" for cloud fraction
endif
...
...

```
Please let me know your thoughts.

Comment (hu5970) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: Do we know if cldfra and cldfra_bl will not coexist?  
If so, the check can be:
if (iret /= nf90_noerr .and. trim(varname) == "cldfra") then

Thanks,
Ming

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: @ericaligo-NOAA, @ChunxiZhang-NOAA Could you answer Ming's question above?

Comment (SamuelTrahanNOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: Isn't that determined by what you put in your diag table?

Comment (SamuelTrahanNOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: I added logic to automatically choose cldfra or cldfra_bl. Could everyone please try the code and make sure it works for your configuration?

It'll use whichever variable is present. If both are present, it'll give a warning and use cldfra.

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: @SamuelTrahanNOAA Your solution looks great to me.  I will test with RRFS model history files including cldfra. Thanks!

Comment (SamuelTrahanNOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: @ericaligo-NOAA @hu5970 Could one of you please test this in the RRFS workflow?

Comment (EricJames-NOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: @SamuelTrahanNOAA I will test it and make sure it works!  Thanks.

Comment (EricJames-NOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: @SamuelTrahanNOAA it works fine for RRFS!  Thanks for the quick coding work.

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA From the UPP RT tests on Hera, there are following changed results in rap/hrrr:
```
[Wen.Meng@hfe07 ~/stmp2]$ cat rap_2020072316/WRFPRS.GrbF16.diff
697:190713718:HGT:cloud base:rpn_corr=0.895998:rpn_rms=3364.31
698:191867295:PRES:cloud base:rpn_corr=0.91432:rpn_rms=15296.7
708:195473891:CDCON:convective cloud layer:rpn_corr=-nan:rpn_rms=undefined
[Wen.Meng@hfe07 ~/stmp2]$ cat rap_2020072316/WRFNAT.GrbF16.diff
1070:282595626:HGT:cloud base:rpn_corr=0.895998:rpn_rms=3364.31
1071:283749203:PRES:cloud base:rpn_corr=0.91432:rpn_rms=15296.7
[Wen.Meng@hfe07 ~/stmp2]$ cat hrrr_2020060118/WRFTWO.GrbF04.diff
121:85620587:HGT:cloud base:rpn_corr=0.932971:rpn_rms=2896.63
122:88068643:PRES:cloud base:rpn_corr=0.931465:rpn_rms=12453.8
```
I would assume these changes come from updates in CLDRAD.f?

Now I saw valid cloud fraction in GSL RRFS test at /scratch1/NCEPDEV/stmp2/Wen.Meng/post_rrfs_2022102713.
Would you please verify my testing? Thanks!

**Comment by WenMeng-NOAA** [comment]
@SamuelTrahanNOAA Would you add change logs in two source code files CLDRAD.f and INITPOST_NETCDF.f?

**Comment by EricJames-NOAA** [comment]
> @EricJames-NOAA From the UPP RT tests on Hera, there are following changed results in rap/hrrr:
> 
> ```
> [Wen.Meng@hfe07 ~/stmp2]$ cat rap_2020072316/WRFPRS.GrbF16.diff
> 697:190713718:HGT:cloud base:rpn_corr=0.895998:rpn_rms=3364.31
> 698:191867295:PRES:cloud base:rpn_corr=0.91432:rpn_rms=15296.7
> 708:195473891:CDCON:convective cloud layer:rpn_corr=-nan:rpn_rms=undefined
> [Wen.Meng@hfe07 ~/stmp2]$ cat rap_2020072316/WRFNAT.GrbF16.diff
> 1070:282595626:HGT:cloud base:rpn_corr=0.895998:rpn_rms=3364.31
> 1071:283749203:PRES:cloud base:rpn_corr=0.91432:rpn_rms=15296.7
> [Wen.Meng@hfe07 ~/stmp2]$ cat hrrr_2020060118/WRFTWO.GrbF04.diff
> 121:85620587:HGT:cloud base:rpn_corr=0.932971:rpn_rms=2896.63
> 122:88068643:PRES:cloud base:rpn_corr=0.931465:rpn_rms=12453.8
> ```
> 
> I would assume these changes come from updates in CLDRAD.f?
> 
> Now I saw valid cloud fraction in GSL RRFS test at /scratch1/NCEPDEV/stmp2/Wen.Meng/post_rrfs_2022102713. Would you please verify my testing? Thanks!

@WenMeng-NOAA yes, we changed the formulation of the exp2 ceiling diagnostic, which would affect these fields.  We are seeking to improve the diagnostic for RRFS for the FAA.  Your GSL RRFS test looks good to me!  Thanks!

**Comment by WenMeng-NOAA** [comment]
The UPP RT tests were completed on WCOSS2, Hera and Orion. The updates of  baselines for rap and hrrr are needed with this PR.

**Comment by WenMeng-NOAA** [comment]
@SamuelTrahanNOAA Would you add change logs in two source code files CLDRAD.f and INITPOST_NETCDF.f so I can proceed to merging?

**Comment by SamuelTrahanNOAA** [comment]
> Would you add change logs in two source code files CLDRAD.f and INITPOST_NETCDF.f so I can proceed to merging?

It is done.

**Review by WenMeng-NOAA — APPROVED** [review]

**Pull request merged**